drafts/two-way-binding-signals-and-query-params.md

```diff
@@ -0,0 +1,754 @@
+
+When we look at this setup with the two effects, it seems like is going to enter in an infinite loop: the first effect makes `activatedRouteValue` to update `queryParamValue`, this will trigger the second effect, where `queryParamValue` will update `activatedRouteValue`, which will trigger the first effect again, and so on and so forth.
+
+This infinite loop would happen if the values that are being assigned are different every time. That's one of the reasons we have to be very careful when we use `allowSignalWrites: true`. However, in this case we are safe. To illustrate it, let's consider the following situation:
+
+```typescript
+const activatedRouteValue = signal("red");
+const queryParamValue = signal("blue");
+
+// Something triggers a change in one signal. For example:
+activatedRouteValue.set("purple");
+
+// activatedRouteValue has changed from "red" to "purple" then trigger first effect:
+
+// (Inside activatedRouteValue->queryParamValue effect):
+queryParamValue.set("purple");
+
+// queryParamValue has changed from "blue" to "purple" then trigger second effect:
+
+// (Inside queryParamValue->activatedRouteValue effect):
+activatedRouteValue.set("purple");
+
+// activatedRouteValue is already "purple". It hasn't changed its value. Therefore,
+// there's no need to trigger first effect again. End of the chain.
+```
+
+So this should be it, right? Well, unfortunately this doesn't work yet:
+
+![Empty UI situation when using our almost ready solution](https://cdn.hashnode.com/res/hashnode/image/upload/v1710538597922/uDBmH5mo6.gif?auto=format)
+
+Apparently we have an empty state. Not only that, query params disappear from the browser url. What is going on?
+
+Well, most of it is related to Angular's router start up logic. Initially, Angular only emits the actual or real values of the query params through the `activatedRoute.queryParams` observable **when the first navigation** is completed. 
+
+This sounds surprising at first, specially when we try to access a query param that is initially provided in the browser url and, through Angular's router API, its starting value is `undefined`. Why doesn't Angular just pick whatever is initially in the browser url?
+
+To understand it better let's consider an example. Imagine that you visit `http://localhost:4200/home?firstName=Dave` and, in the component assigned to `/home`, you try to access `activatedRoute.snapshot.queryParams["firstName"]`, in its `ngOnInit` for example. Angular won't return `Dave` as you may expect. You will actually receive `undefined`. The reason is that Angular hasn't finished the initial navigation yet. During this period of time, if we access the `snapshot` or the current value of the `activatedRoute.queryParams` observable, we will get an empty object `{}`, that's why we received `undefined` when we try to access an specific query param during start up.
```

Comment:
**thought:** this paragraph mostly rehashes the previous two, maybe you can merge the last three paragraphs and make them slightly simpler? This Angular quirk is not necessarily the core interesting part of this article and I don't think we need to go quite that deep into it.